### ISSUE TITLE
Restructure CSS

### ISSUE THREAD

**Issue by justcolin** [projecthydra/sufia#1611]
## Summary

While working on U-M's beta interface for our repository based on Sufia 7 we had a lot of trouble working with Sufia's current stylesheets. The following suggestions are centered around simplifying maintenance and making it easier for groups to customize Sufia's views as much or as little as they see fit.

**Note:** This issue should probably be dealt with as part of implementing whatever the UI working group comes up with instead of working on the current interface.

## Suggestions

1. Reduce the amount of CSS in Sufia to a bare minimum, instead relying on native Bootstrap wherever possible.
  - Bootstrap is a great tool that was made to be easier to customize. If we let Bootstrap do almost all the heavy lifting and trust it to have proper padding/margins it will be a lot easier for developers to customize Sufia to meet their needs
  - This is also reduce the amount of CSS which has to be maintained by the Sufia community.
2. Any custom CSS shouldn't have any id's in the selectors, and should only have at most two classes.
  - This will make it easier to overwrite styles and will ease longterm maintenance.
3. Add semantically named classes to most elements in Sufia, even when they aren't being used by stock Sufia's CSS.
  - This will make it easier for people to restyle Sufia without having the go in and edit the HTML.
4. Improve the accessibility of Sufia's stock styles.
  - There are a number of things that need improvement so that it is easier for more people to use Sufia, such as more obvious link styles, clear `:hover` and `:focus` actions, etc.

## Related
See #1594 for the start of this discussion
See #1608 for the discussion of front-end development best practices

**Comment by justcolin** [comment]
@grosscol

**Comment by jcoyne** [comment]
I think I'm pretty strongly against point one. I think that people want a repository that looks good out of the box.  Not every institution is Michigan, and has the resources to completely redo the CSS.

**Comment by jcoyne** [comment]
I think point 4 got messed up (it's trailing point 3) 

> There are a number of things that need improvement

Please expand on these things.  Let's make all suggestions actionable.

**Comment by grosscol** [comment]
I don't get the impression that many institutions that are not already in the mix will be doing a great deal of work in house.  I imagine that the hydra is a box project will provide a very stylish styling.

For the base Sufia, making easier to modify css should give us a more stable (fewer places to touch) base for UI work.

I do concur with a list of actionable items.

**Comment by jcoyne** [comment]
@grosscol But what about the institutions that *are* already in the mix. Those are the ones I care most about right now.

**Comment by grosscol** [comment]
My assumption was that the existing members have already modified the views to suite their own requirements.  So long as we don't alter the routes and helpers the existing modified views should still function.  Worst case is copying up views from older versions.

**Comment by justcolin** [comment]
On point 1: I think Sufia can be made to look even better out of the box and use less custom CSS at the same time. Blacklight looks really good out of the box, and it uses barely any CSS, mostly just supplementing Bootstrap where needed with some generic classes. Bootstrap is made to look good and contemporary out of the box.

**Comment by awead** [comment]
I'm against point 1 as well because Sufia needs to have a "look" to it, but I agree that it should be simple. I think the other points are good, especially 2 and 3. I think we could create a default Sufia skin that implements all the various bits that are a part of 2 and 3, then the user can either remove that and create their own, or tweak it. This way there's at least a kind of a template for them to work with. I think that balances institutions that have UI resources versus those that don't.

I can't speak to 4 without more information.

**Comment by cam156** [comment]
@justcolin @jcoyne I wonder if we could accomplish more customization in the css by generating the css into the app instead of hiding it within sufia.  I think it would be great to use the two perspectives to allow for customization to be done more easily in Sufia!

**Comment by awead** [comment]
@cam156 yes, that's what I was thinking about with my "skin" comment. The generated CSS would use the classes and selectors from points 2 and 3.

**Comment by justcolin** [comment]
@cam156 I like the idea of generating the CSS into the app, or providing a gem which creates the skinned version, as long as the unskinned version still works and looks acceptable.

**Comment by awead** [comment]
@justcolin I think the point of the generated CSS is that it would be the default look. There wouldn't be an "unskinned" version. Otherwise, you're maintaining two different looks, the default skin and the unskinned. Just have one default look that's generated into the application and can be overridden easily. It also goes to what @jcoyne and I were saying that Sufia needs a polished look out of the box.

**Comment by justcolin** [comment]
@awead I am not so sure about that. If we go with the "use Bootstrap as much as possible for layout" idea it wouldn't be any more work to have Sufia look okay without that generated CSS. If a couple of things are broken, that's fine, but encouraging the site to be mostly passable without the generated CSS would encourage better use of Bootstrap to get most of the work done. That would also the go towards simplifying the generated CSS.

**Comment by awead** [comment]
Again, this goes back to point 1: Sufia needs to look good out of the box. To do that, we need to test the complete application, including its generated CSS, otherwise we don't know the quality of the finished product. The CSS can be generated as part of the install process, but can be overridden. If the user wants to delete that generated file and use the "skinless" version, they're free to do that, but I don't see the point in having to support that scenario in our automated testing. The generated CSS can be as simple as possible, but it needs to be there, in my mind, because that's what the full application is.

**Comment by jcoyne** [comment]
If you want an unskinned version, can't you just remove the generated sufia lines from application.css in your local install?

**Comment by justcolin** [comment]
@awead: I believe we are in agreement. I don't think the automated testing should be tested without the the generated CSS. However, I do think that better practices would be encouraged if while working on the next UI we tried to get as far as we can without using any generated CSS. It would help everyone have an easier time.

@jcoyne: You can, but it destroys a lot of the layout, causing one to have to do a lot of extra work. It is possible to use Bootstrap more for doing layout work, reduce the dependency on custom CSS, while still having a fantastic looking website and making it easier for people trying to customize Sufia in small or large amounts.

**Comment by justcolin** [comment]
@jcoyne: My goal in these suggestions isn't to make things worse for people using Sufia out of the box while making it easier for people who want to customize the website. I want to make sure both parties have a better experience.

**Comment by awead** [comment]
So now I'm asking myself: what does done look like? Currently, Sufia's "skin" is:
https://github.com/projecthydra/sufia/blob/master/app/assets/stylesheets/sufia.css.scss
Which, in looking at, doesn't look too bad to me. The only thing I'd like to see is what I mentioned in the TODO... use import exclusively and not require. Or have the requires somewhere else so it's explicit what parts are using asset pipeline and what parts are using Sass.

It seems to me we're talking about restructuring parts of the imported css (settings, header, styles, etc.) and perhaps floating them up into the sufia.css file where they can be overridden there. However, the current strategy still looks pretty good to me: I simply add one line to my application.css file and be done. Or, I can copy sufia.css into my project and make selective overrides directly in that file, or I can make more large-scale changes by not importing some of the files and substituting my own.

So I guess what I'm getting at is, what don't we like about the current process, and what are we trying to change about it?

**Comment by justcolin** [comment]
@awead If all SASS files are partials imported into sufis.css.scss then that is good enough (that wasn't part of the original issue). See the **Suggestions** section of this issue for my main suggestions.

**Comment by justcolin** [comment]
@jcoyne: I strongly believe that more use of Bootstrap will not make Sufia look worse out of the box. Arguably it will make it look a lot better.

If Bootstrap is used for all or almost all of the layout we will end up with a robust, easier to maintain design. Then we can overwrite some of Boostrap's reset to customize text and/or UI elements to keep compatibility with Bootstraps classes, but make it our own. Finally, with a few custom rules to add a little more branding you can end up with a distinctive, modern, robust design that looks great, and has less code making it easier to maintain & alter.

**Comment by jcoyne** [comment]
@justcolin I'll keep an open mind while I eagerly await a PR.

**Comment by justcolin** [comment]
@jcoyne I assume you don't want any pull requests for the current interface, correct (since large portions are about to be rewritten anyways)?

**Comment by jcoyne** [comment]
As far as I know, only the upload / create a new work is changing.  I'm not the person who is most involved in the Sufia plan though.

**Comment by justcolin** [comment]
Okay; once the wireframes are set and implementation planning begins I will investigate updating the pages that aren't being touched.

**Comment by mjgiarlo** [comment]
@justcolin Yes, it may be easier to work on this once all the UI-related issues on the 7.0.0 release are completed.

**Comment by mtribone** [comment]
@justcolin @jcoyne @mjgiarlo The UI working group is proposing add and edit new work, new work show page, a new top bar, and may touch things like the tabbed my works, my collections, etc. and the dashboard.

**Comment by jcoyne** [comment]
@mtribone What's the projected timeline for completing that work?

**Comment by mtribone** [comment]
@jcoyne We have most of the issues in Github now prefaced with UI. Anyone can pick up the following issues and get crackin' and/or give us feedback and/or ask questions. Right now we are missing the show page issue because we are still discussing the layout and hierachy.

https://github.com/projecthydra/sufia/issues/1562
https://github.com/projecthydra/sufia/issues/1563
https://github.com/projecthydra/sufia/issues/1572
https://github.com/projecthydra/sufia/issues/1573
https://github.com/projecthydra/sufia/issues/1577
https://github.com/projecthydra/sufia/issues/1578
https://github.com/projecthydra/sufia/issues/1583
https://github.com/projecthydra/sufia/issues/1591

**Comment by jcoyne** [comment]
@mtribone :+1:

**Comment by mjgiarlo** [comment]
@justcolin Given that a Hydra Dev Congress event in Ann Arbor was recently announced, I wonder whether some of the work that you ticketed here might be good candidates for a multi-person face-to-face hack in May. The timing on that would be good, I believe, and that way there's a captive audience and there are likely to be multiple Sufia-using folks in the room.

**Comment by awead** [comment]
@mjgiarlo good idea... maybe we can send @mtribone as our delegate, provided the PA legislature gets their act together.

**Comment by mjgiarlo** [comment]
@awead :+1: 

@justcolin Might that timing work for your purposes too?

**Comment by justcolin** [comment]
@mjgiarlo: That could work well for us, I'll just have to check with the higher ups.

**Comment by mtribone** [comment]
@mjgiarlo Yeah that would fantastic to use the upcoming dev meeting in May! But yes, sadly we have no idea when the travel ban will be lifted. However, this work is operationally critical...

**Comment by justcolin** [comment]
@mjgiarlo, @mtribone and @jcoyne: Some of the issues we encountered were related to odd or overly specific styles in Curation Concerns. Would you all be open to remove the Curation Concerns CSS from Sufia so they are decoupled, or would you rather fix the CSS in Curation Concerns as well?

There isn't much, but it is just enough to provide some issue.

**Comment by jcoyne** [comment]
I'm a bit surprised that you are using any CurationConcerns styles in sufia. I would think that those aren't required.

**Comment by justcolin** [comment]
Scratch that, I was just looking at it again and it looks like Bootstrap's is being included in a file called Curation Concerns when you load the site in dev mode. The styles we were fighting were just unusual uses of Bootstrap.

**Comment by justcolin** [comment]
@jcoyne: What's the reasoning for closing this ticket? I am not opposed to it, just wondering what the reasoning is.

**Comment by mjgiarlo** [comment]
I'm guessing it was a misfire. It's re-opened now, @justcolin.

**Comment by jcoyne** [comment]
Oops. I must have hit it by mistake

**Comment by mjgiarlo** [comment]
GitHub makes it too easy to click "Close and comment!"

**Comment by justcolin** [comment]
It must be to promote faster development. Doesn't matter if the work actually gets done, just matters if tickets are getting closed. :-P

**Comment by mjgiarlo** [comment]
Ha! #worksforme #wontfix ;)

**Comment by justcolin** [comment]
I suggest that we use BEM for the CSS restructure. See [this comment](https://github.com/projecthydra/sufia/issues/1608#issuecomment-193448811) for the reasoning, and keep the discussing on #1608.

**Comment by mtribone** [comment]
@justcolin Sounds good. Yeah, I've been researching BEM a little bit and agree that its syntax is a little ugly. However, if it will do the job and help other people fix/customize the code I'm all for it. I might start playing around with it locally and refactor some views.

**Comment by mjgiarlo** [comment]
@mtribone @justcolin @grosscol @awead @cam156 See the PRs above. This work is starting to happen in bite-sized chunks.

**Comment by mtribone** [comment]
@mjgiarlo Yeah I see that happening. I was experimenting with BEM and the masthead, but @cbeer is refactoring a lot of stuff. So Imma gonna wait and work on something else in the meantime while things shake out. Are we still interested in using BEM on top of Bootstrap?

**Comment by mjgiarlo** [comment]
@mtribone What do you think, @cbeer @jcoyne? I'm not familiar enough with BEM to comment at the moment. It would be good to come up with a plan so we don't step on each other's toes!

**Comment by cbeer** [comment]
I'd guess doing BEM will be easier if we clean things up first.

**Comment by justcolin** [comment]
@mjgiarlo, @mjgiarlo, @jcoyne and @cbeer: Even if we only use Bootstrap and manage to remove all the custom CSS, adding BEM style classes to everything would still be a huge help. It would make customization easier for everyone (no need to edit templates if a class is already there), and if we ever need to add custom CSS in the future we will already have a class there, ready and waiting.

If you don't use BEM, that's fine, it just provides a defined and easy to understand way to name classes. The most important thing to me is to remove all id's for styling (id's for JS are fine) and give almost everything a class instead.

**Comment by mtribone** [comment]
@mjgiarlo Bottstrap 4 is going from Less to Sass. One thing that I've experienced so far is trying to override specificity and had to follow Bootstrap syntax over BEM. For example:

https://github.com/projecthydra/sufia/blob/masthead_header/nav/app/assets/stylesheets/sufia/_header.scss#L48

@cbeer Yeah, man!

**Comment by atz** [comment]
This ticket is too big to be useful to developers.  Nice conversation and all, but I would recommend piecing it out and closing this one.

**Comment by mjgiarlo** [comment]
This one is for discussion only. I believe @justcolin was going to take a gander at the UI changes we've been making the past few weeks (and possibly submit a PR). We may take up this work at the May Hydra Dev Congress in Ann Arbor. I'd like to leave it open for now, and leave it on the backlog, until we've ticketized the work here or punted it.

**Comment by justcolin** [comment]
@mjgiarlo: I've been looking at the UI changes and have some comments, but since you all are moving so fast pushing 7.0 out the door I haven't really have the chance to get any comments or PR's in. I am waiting until the Dev Congress or after 7.0 is out the door.

**Comment by mjgiarlo** [comment]
Sounds like a plan. At a macro level, would you say the changes you've seen are a step in the right direction or digging a deeper hole?

**Comment by justcolin** [comment]
There is a good amount of positive movement going on,  and some lateral movement which is understandable. My primary concerns currently are sustainability and the designs, but that is a longer conversation.

**Comment by mjgiarlo** [comment]
@justcolin @mtribone As we start to clarify the paths forward for Sufia and Hyrax, I want to make sure we don't lose the good discussions here. Are there any bits from the above discussion that we should write up as actionable issues?

**Comment by mtribone** [comment]
Keep specificity low where we can, but overriding Bootstrap makes for very specific code. Use classes over IDs to gradually add granular styling. For example `.btn` `.btn-alert` `.btn-lg`

I've also been using this guide as a reference:
http://codeguide.co/#css

**Comment by justcolin** [comment]
See [this commen](https://github.com/projecthydra/sufia/issues/1608#issuecomment-268597039)t for a suggestion on how we could make it easier to make this issue actionable.

**Comment by mjgiarlo** [comment]
@justcolin Looks good.

@mtribone Would documenting that or similar practices in our Dev Guide (or elsewhere in our docs) be sufficient?

**Comment by mtribone** [comment]
@mjgiarlo Yeah the Dev Guide is a good idea!

**Comment by mjgiarlo** [comment]
@justcolin @mtribone are either of you inclined to start building out a new section of the dev guide about this?

**Comment by mtribone** [comment]
@mjgiarlo I certainly don't mind, but it might be a bit of a slow burn while we crank out ScholarSphere. Are we at a point with Sufia that it is default enough?

**Comment by justcolin** [comment]
@mjgiarlo I can contribute to it and may be a bit after since I am not currently working on any Sufia projects.

**Comment by mtribone** [comment]
@justcolin You have a good chunk of recommendations here: https://github.com/projecthydra/sufia/issues/1608#issue-137125284

**Comment by mjgiarlo** [comment]
@justcolin @mtribone Sounds good, y'all. Once you add the new recommendations to the wiki, hit me up and I'll be happy to review. Feel free to close this issue when that's ready. Thanks!

**Comment by mjgiarlo** [comment]
@justcolin @mtribone FWIW, the folks at UW have been keeping some somewhat related notes on their own wiki. I link that here in case there are interesting tidbits to repurpose for the Sufia Dev Guide: https://bitbucket.org/uwlib/druw/wiki/User%20Interface%20Customization%20in%20Sufia.md 

HT: @metageeky & @younga3

**Comment by mtribone** [comment]
@mjgiarlo The work from UW looks pretty good. I think we have several things to outline, if I'm following the conversations correctly

1. CSS Guidelines for writing CSS per @justcolin's suggestions and perhaps https://github.com/projecthydra/sufia/issues/1608#issuecomment-260655039
1. Overriding styles and customization using @metageeky & @younga3 work as a starting point
1. CSS Linter per @justcolin 
1.  Style Guide per @jenlindner https://github.com/projecthydra/sufia/issues/1608#issuecomment-250215268

**Comment by justcolin** [comment]
#3 is under way, it will just take a bit to tweak the rules and to fix up the errors it throws for existing CSS. As for #1, I should be able to get started on it soon.

**Comment by mjgiarlo** [comment]
@mtribone @justcolin Solid! Thanks, you two.
